packages/siwe-parser/lib/siwe-abnf.txt

```diff
@@ -0,0 +1,210 @@
+;    This will simiplify recognizing contexts AND remove unneccesary callbacks
+; 2.a) domain is defined as authority-d which is identical to authority except that there will be no
+;      callback functions defined on authority-d or any of its *-d components.
+; 2.b) The resource URI is defined as URI-r and its components defined as *-r.
+;      In this way, callback functions can be defined on URI and is components while
+;      leaving URI-r to be parsed identically with no unnecessary callback functions to slow it down.
+; 3) IPv6address does not work because of APG's "first-success disambiguation" and "greedy" repetitions.
+;    IPv6address redefined and validations moved to callback functions (semantic vs syntactic validation)
+;    Redefinition requires negative look-ahead operators, https://en.wikipedia.org/wiki/Syntactic_predicate
+;    That is SABNF instead of simple ABNF.
+; 4) IPv4address fails because of "first-success disambiguation".
+;    This could be fixed with rearrangement of the alternative terms. However, it would still not
+;    accept zero-padded (leading zeros) decimal octets.
+;    Therefore, IPv4address is also done with callback functions and semantic validation.
+; 5) The negative look-ahead operator is also needed in the definition of host to
+;    prevent failure with a reg-name that begins with an IPv4 address.
+; 6) NOTE: host = 1.1.1.256 is a valid host name even though it is an invalid IPv4address.
+;          The IPv4address alternative fails but the reg-name alternative succeeds.
+
+
+sign-in-with-ethereum =
+    oscheme domain %s" wants you to sign in with your Ethereum account:" LF
+    address LF
+    LF
+    [ statement LF ]
+    LF
+    %s"URI: " URI LF
+    %s"Version: " version LF
+    %s"Chain ID: " chain-id LF
+    %s"Nonce: " nonce LF
+    %s"Issued At: " issued-at
+    [ LF ex-title expiration-time ]
+    [ LF nb-title not-before ]
+    [ LF ri-title request-id ]
+    [ LF re-title resources ]
+ex-title        = %s"Expiration Time: "
+nb-title        = %s"Not Before: "
+ri-title        = %s"Request ID: "
+re-title        = %s"Resources:"
+oscheme         = [ ALPHA *( ALPHA / DIGIT / %d43 / %d45-46 ) "://" ]
+domain          = authority-d
+address         = "0x" 40*40HEXDIG
+    ; Must also conform to captilization
+    ; checksum encoding specified in EIP-55
+    ; where applicable (EOAs).
+
+statement       = 1*( %d97-122 / %d65-90 / %d48-57 / %d32-33 / %d35-36 / %d38-59 / %d61 / %d63-64 / %d91 / %d93 / %d95 / %d126)
+    ; The purpose is to exclude LF (line breaks).
+    ; LDT 10/04/2023: Do you mean %d32-126? All printing characters
+version         = "1"
+nonce           = 8*( ALPHA / DIGIT )
+issued-at       = date-time
+expiration-time = date-time
+not-before      = date-time
+request-id      = *pchar
+chain-id        = 1*DIGIT
+    ; See EIP-155 for valid CHAIN_IDs.
+resources       = *( LF resource )
```

Comment:
So this PR closes #30 then?

Reply:
At first glance, I think so. But I'll need a little more time to go through #30 in more detail for a definitive reply.

Reply:
Yes, issue #30 can be closed now. The `siwe-parser` matches the spec for both `statement` and `resources`. I should have read #30 more closely. The spec ABNF grammar format does indeed allow for present, empty or missing for both the `statement` and the `resources`. With the latest commit, the `siwe-parser` matches the spec for both `statement` and `resources`. There are unit tests for all cases in `t-spec.test.ts`.